[Load] Centrifugal and Coriolis rotation forces

aquinorayme opened the pull request:
**Describe the PR**
This PR implements centrifugal and Coriolis rotation forces to simulate rotating reference frames without rotating the mesh. The main features are given as follows.

The solver can now apply rotational body forces based on a defined origin and angular velocity. The input file requires a rotation_forces definition inside external_loading_conditions to specify the origin coordinates, omega, and rotation direction.

A protected rotation block is created in MPMBase to store these variables. The MPMExplicit solver is modified to check for this rotation block and delegate the computation to the scheme. The MPMScheme class is updated with compute_rotation_forces, which iterates over particles and passes the rotation arguments.

The corresponding particle function map_rotation_force is declared in ParticleBase as a pure virtual function. The Particle class implements the formulation for the Centrifugal and Coriolis forces and maps them to the nodes.

Finished and following:

- Centrifugal and Coriolis force formulation in Particle
- Input parsing in MPMBase
- Updates to MPMExplicit and MPMScheme to pass rotation arguments
- Pure virtual interface in ParticleBase
- Updated unit tests (mpm_scheme_test.cc) for compatibility
- Applied clang-format

**Related Issues/PRs**
N/A

**Additional context**
N/A

aquinorayme committed:
added rotation force input in mpm_base

aquinorayme committed:
Passed rotation parameters to compute_forces

aquinorayme committed:
Defined compute_forces to accept and apply rotation arguments

aquinorayme committed:
Implemented centrifugal and Coriolis force calculations

aquinorayme committed:
Updated scheme tests to include dummy rotation arguments

aquinorayme commented:
I think the failing test should just be an issue with formatting? I will run the formatting command again and see if that fixes things.

aquinorayme committed:
Applied clang-format fixes

aquinorayme committed:
Fixed formatting according to cppcheck

bodhinandach commented on include/particles/particle.h:
```diff
@@ -237,6 +237,11 @@ class Particle : public ParticleBase<Tdim> {
   void map_body_force(const VectorDim& pgravity) noexcept override;
 
+  //! Map rotation forces (centrifugal and coriolis)
+  void map_rotation_force(const VectorDim& rotation_origin,
```
Please add short descriptions on the input parameters, see above.

adrienmartinezZ replied:
Done!

bodhinandach commented on include/particles/particle.tcc:
```diff
@@ -1728,3 +1728,35 @@ inline double mpm::Particle<Tdim>::compute_asflip_beta(double dt) noexcept {
+  const VectorDim f_total =
+      (-mass_ * (omega_vec_3d.cross(omega_vec_3d.cross(r_p_3d)) +
+                 2.0 * omega_vec_3d.cross(vel_3d)))
+          .head<Tdim>();
```
Why f_total is VectorDim not  Eigen::Matrix<double, 3, 1>? All the other components are Eigen::Matrix<double, 3, 1>.

bodhinandach commented on include/solvers/mpm_base.h:
```diff
@@ -294,6 +294,13 @@ class MPMBase : public MPM {
   //! Absorbing Boundary Variables
   bool absorbing_boundary_{false};
 
+  bool rotation_forces_{false};
```
Missing description comment like others...

adrienmartinezZ replied:
Done!

bodhinandach reviewed (commented):
Thanks @aquinorayme and @adrienmartinezZ for the initial implementations. I like the initial design. A couple of items that we need to think through before merging this branch:
- How to ramp rotation? Can we add math_function for rotation (and gravity, since we dont have it there too)?
- The current framework only considers the axis of rotation is aligned with the z direction, but works for 2D and 3D. What should be done if axis of rotation is not aligned with the plane of the mesh (arbitrary orientation)?
- How to include this to other scheme, i.e. mpm_implicit?
- Some testing need to be developed before merging.

adrienmartinezZ committed:
Cleaned the format for parameter descriptions.
- Added short descriptions for input parameters in map_rotation_force in particle.h
- Spelled "stress" correctly in the description of initial_stress in particle.h
- Added description comment for rotation_forces_ in mpm_base.h
- Deleted extra "/" in compute forces description in mpm_scheme.h

adrienmartinezZ commented on include/solvers/mpm_scheme/mpm_scheme.h:
```diff
@@ -52,14 +52,20 @@ class MPMScheme {
   virtual inline void pressure_smoothing(unsigned phase);
 
-  //! Compute forces
+  ///! Compute forces
```
Done!

aquinorayme committed:
Applied clang-format fixes

aquinorayme committed:
Fixed Clang formatting errors

aquinorayme committed:
Small test for Rotation Forces v.1

aquinorayme committed:
Applied clang format to v.1

aquinorayme committed:
Added 3D version of json error

aquinorayme committed:
Small test v.2

aquinorayme committed:
Changed map_rotation_forces to rotation_function

aquinorayme committed:
Removed comments

aquinorayme committed:
Updated rotation force tests with boolean pass/fail

aquinorayme committed:
Applied clang format

aquinorayme committed:
Initialize mesh and materials before applying loads in rotation tests

bodhinandach commented on include/particles/particle.tcc:
```diff
@@ -1728,3 +1728,35 @@ inline double mpm::Particle<Tdim>::compute_asflip_beta(double dt) noexcept {
+
+  // Relative Position (r) and Velocity (v) in 3D
+  Eigen::Matrix<double, 3, 1> r_p_3d = Eigen::Matrix<double, 3, 1>::Zero();
+  r_p_3d.head<Tdim>() = this->coordinates_ - rotation_origin;
```
Can we change `head<Tdim>()` to just `head(Tdim)`? Just to unify it with the rest of the coding style.

bodhinandach commented on include/particles/particle.tcc:
```diff
@@ -1728,3 +1728,35 @@ inline double mpm::Particle<Tdim>::compute_asflip_beta(double dt) noexcept {
+
+//! Rotation Function
+template <unsigned Tdim>
+void mpm::Particle<Tdim>::rotation_function(const VectorDim& rotation_origin,
```
I think this should not be renamed... Please rename back to `map_rotation_force()`

aquinorayme replied:
Will do

bodhinandach reviewed (commented):
@aquinorayme Thanks. Can you also add a test in particle_test.cc? See map_body_force() test there.

aquinorayme committed:
Renamed function

aquinorayme committed:
First est in particle_test.cc (v.1)

aquinorayme committed:
Applied clang format for changes

aquinorayme committed:
Added explicit physics validation for centrifugal and coriolis forces

aquinorayme committed:
Flipped signn of expected values

aquinorayme committed:
Attempt v.2

aquinorayme committed:
Removed  comments

aquinorayme committed:
Removed irrelevant comments

aquinorayme committed:
Replaced std::runtime error

aquinorayme committed:
Removed this-> as discussed, added "using", and applied clang format

aquinorayme committed:
Fixed mpm_explicit.h removed line

aquinorayme committed:
Applied clang format (v.3)

aquinorayme commented on include/solvers/mpm_base.tcc:
```diff
@@ -790,6 +790,30 @@ void mpm::MPMBase<Tdim>::initialise_loads() {
+    console_->error(
+        "#{}: Rotation forces are undefined; Rotation forces JSON data "
+        "not found",
+        __LINE__);
```
I wrote this above,
throw std::runtime_error(
          "mpm::base::initialise_loads(): Specified rotation origin dimension "
          "is invalid");
and then wrote the else statement, 
  } else {
    console_->error(
        "#{}: Rotation forces are undefined; Rotation forces JSON data "
        "not found",
        __LINE__);
I used error instead of warn since you mentioned there were too many warns already.

I followed the convention,
      if (!particles_tractions)
        throw std::runtime_error(
            "mpm::base::initialise_loads(): Particles tractions are not "
            "properly assigned");
    }
  } else
    console_->warn(
        "#{}: Particle surface tractions are undefined; Particle surface "
        "tractions JSON data not found",
        __LINE__);

bodhinandach replied:
I see... but this should not be an error. It should be a warn.

bodhinandach replied:
Thanks, resolving this.

aquinorayme committed:
Changed to warn instead of error

adrienmartinezZ committed:
Added gravity and rotational forces ramping
-Added ramping_time parameter to mpm_base.h
-Implemented ramping logic in mpm_base.tcc "initiatlise_loads()" with negative throws + same ramping throws
-Apply ramping multipliers to gravity and rotation forces in mpm_explicit.tcc "update_particle_velocities()"
-Initialized write_json_ramping() function to write_mesh_particles.h and write_mesh_particles.cc for test
-Add ramping test cases to mpm_explicit_usf_test.cc

Test Cases for Ramping:
- Case 1: Ramping gravity with t > 0s
- Case 2: Ramping gravity with t < 0s (error case 1)
- Case 3: Ramping rotation with t > 0s
- Case 4: Ramping rotation with t < 0s (error case 2)
- Case 5: Ramping both gravity and rotation at same time (error case 3)
- Case 6: No ramping (error case 4)

adrienmartinezZ committed:
Added 3D case to test ramping
-3D case to mpm_explicit_usf_test.cc for gravity ramping

adrienmartinezZ committed:
Changing 3D case to solve() to run.

adrienmartinezZ committed:
Added tests to cover code and more warnings to check.

adrienmartinezZ committed:
Reverting warnings tests to cover branches for 2D and 3D cases.

bodhinandach committed:
Merge branch 'master' of https://github.com/geomechanics/mpm into load/centrifugal

bodhinandach committed:
Add json search

bodhinandach committed:
modifications including two different ramping times for gravity and rotation force

adrienmartinezZ committed:
Independent multipliers for gravity and rotational ramping
Testing for 2D and 3D cases with the following combinations:
-Gravity only
-Rotational ramping only
-Both gravity and rotational ramping

adrienmartinezZ committed:
Fixed formatting for clang-tidy issues in mpm_base.tcc

aquinorayme committed:
Added tests to cover IO missing file error and time-ramping multiplier cutting off

aquinorayme committed:
Updated explicit solver Catch2 tests to execute a full solve() cycle instead of only initialise_loads()

aquinorayme committed:
Refactored ramping JSON generation to execute full solve loops

aquinorayme committed:
Refactored ramping JSON generation to fix exception coverage

aquinorayme committed:
Refactored ramping JSON generation to fix exception coverage

aquinorayme committed:
Added 3D clamping tests and executed full solve loops

aquinorayme committed:
Fixed final coverage gaps for base warnings, IO search, and Eigen operations

aquinorayme committed:
Combined redundant ramping and clamping test cases to reduce lines of code

bodhinandach committed:
Modify author contribution names

bodhinandach reviewed (commented):
Good job @adrienmartinezZ and @aquinorayme. I am OK with merging this.
